microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/properties/PropertiesFileTextDocumentService.java

```diff
@@ -112,48 +122,46 @@ public void didSave(DidSaveTextDocumentParams params) {
-			if (projectInfo.getProperties().isEmpty()) {
-				return CompletableFuture.completedFuture(Either.forRight(new CompletionList()));
+		return getPropertiesModel(params.getTextDocument(), (document, cancelChecker) -> {
+			// Get MicroProfile project information which stores all available MicroProfile
+			// properties
+			MicroProfileProjectInfoParams projectInfoParams = createProjectInfoParams(params.getTextDocument());
+			MicroProfileProjectInfo projectInfo = getProjectInfoCache().getProjectInfo(projectInfoParams).getNow(null);
+			if (projectInfo == null || projectInfo.getProperties().isEmpty()) {
+				return Either.forRight(new CompletionList());
 			}
-			// then get the Properties model document
-			return getPropertiesModel(params.getTextDocument(), (cancelChecker, document) -> {
-				// then return completion by using the MicroProfile project information and the
-				// Properties model document
-				CompletionList list = getPropertiesFileLanguageService().doComplete(document, params.getPosition(),
-						projectInfo, sharedSettings.getCompletionCapabilities(), sharedSettings.getFormattingSettings(),
-						null);
-				return Either.forRight(list);
-			});
+			cancelChecker.checkCanceled();
+
+			// then return completion by using the MicroProfile project information and the
+			// Properties model document
+			CompletionList list = getPropertiesFileLanguageService().doComplete(document, params.getPosition(),
+					projectInfo, sharedSettings.getCompletionCapabilities(), sharedSettings.getFormattingSettings(),
+					cancelChecker);
+			return Either.forRight(list);
 		});
 	}
 
 	@Override
 	public CompletableFuture<Hover> hover(HoverParams params) {
-		// Get MicroProfile project information which stores all available MicroProfile
-		// properties
-		MicroProfileProjectInfoParams projectInfoParams = createProjectInfoParams(params.getTextDocument());
-		return getProjectInfoCache().getProjectInfo(projectInfoParams).thenComposeAsync(projectInfo -> {
-			if (projectInfo.getProperties().isEmpty()) {
-				return CompletableFuture.completedFuture(null);
-			}
-			// then get the Properties model document
-			return getPropertiesModel(params.getTextDocument(), (cancelChecker, document) -> {
-				// then return hover by using the MicroProfile project information and the
-				// Properties model document
-				return getPropertiesFileLanguageService().doHover(document, params.getPosition(), projectInfo,
-						sharedSettings.getHoverSettings());
-			});
+		return getPropertiesModel(params.getTextDocument(), (document, cancelChecker) -> {
+			// Get MicroProfile project information which stores all available MicroProfile
+			// properties
+			// Don't block if it hasn't been computed yet
+			MicroProfileProjectInfoParams projectInfoParams = createProjectInfoParams(params.getTextDocument());
+			MicroProfileProjectInfo projectInfo = getProjectInfoCache().getProjectInfo(projectInfoParams).getNow(null);
+			cancelChecker.checkCanceled();
```

Comment:
I have some NPE when I hover a property key while loading of the project info.

```
juin 15, 2022 3:39:21 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4mp.commons.MicroProfileProjectInfo.getProperties()" because "info" is null
java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4mp.commons.MicroProfileProjectInfo.getProperties()" because "info" is null
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4mp.commons.MicroProfileProjectInfo.getProperties()" because "info" is null
	at org.eclipse.lsp4mp.utils.PropertiesFileUtils.getProperty(PropertiesFileUtils.java:184)
	at org.eclipse.lsp4mp.services.properties.PropertiesFileHover.getPropertyKeyHover(PropertiesFileHover.java:152)
	at org.eclipse.lsp4mp.services.properties.PropertiesFileHover.doHover(PropertiesFileHover.java:102)
	at org.eclipse.lsp4mp.services.properties.PropertiesFileLanguageService.doHover(PropertiesFileLanguageService.java:104)
	at org.eclipse.lsp4mp.ls.properties.PropertiesFileTextDocumentService.lambda$hover$3(PropertiesFileTextDocumentService.java:156)
	at org.eclipse.lsp4mp.ls.commons.ModelTextDocuments.lambda$computeModelAsync$0(ModelTextDocuments.java:119)
	... 7 more

```

After seeing the code of the hover, it seems projectInfo is required. To avid those NPE please check that projectInfo is null  and in this case don't consume the doHover of the language service.

Reply:
Oops. Should be fixed now